firebase-segmentation/src/main/java/com/google/firebase/segmentation/FirebaseSegmentation.java

```diff
@@ -107,70 +113,88 @@ public static FirebaseSegmentation getInstance(@NonNull FirebaseApp app) {
    * </pre>
    */
   private Task<Void> updateCustomInstallationId(String customInstallationId) {
-    CustomInstallationIdCacheEntryValue cacheEntryValue = localCache.readCacheEntryValue();
-    if (cacheEntryValue != null
-        && cacheEntryValue.getCustomInstallationId().equals(customInstallationId)
-        && cacheEntryValue.getCacheStatus() == CustomInstallationIdCache.CacheStatus.SYNCED) {
-      // If the given custom installation id matches up the cached
-      // value, there's no need to update.
-      return Tasks.forResult(null);
-    }
+    TaskCompletionSource<Void> taskCompletionSource = new TaskCompletionSource<>();
+
+    executor.execute(
```

Comment:
2 points:
* There is a dedicated `Tasks.call(executor, callable)` to achieve this effect
* Consider extracting the lambda below into its own (private) method that you can annotate with `@WorkerThread`

This applies to the here and throughout as well

Reply:
Very good learning!
Done